crypto/initthread.c

```diff
@@ -83,64 +84,100 @@ static GLOBAL_TEVENT_REGISTER *get_global_tevent_register(void)
 static void init_thread_remove_handlers(THREAD_EVENT_HANDLER **handsin);
 static void init_thread_destructor(void *hands);
 static int  init_thread_deregister(void *arg, int all);
 #endif
 static void init_thread_stop(void *arg, THREAD_EVENT_HANDLER **hands);
 
+static THREAD_EVENT_HANDLER ** get_thread_event_handler(OSSL_LIB_CTX *ctx)
+{
+#ifdef FIPS_MODULE
+    return CRYPTO_THREAD_get_local_ex(CRYPTO_THREAD_LOCAL_TEVENT_KEY, ctx);
+#else
+    if (destructor_key.sane != -1)
+        return CRYPTO_THREAD_get_local(&destructor_key.value);
+    return NULL;
+#endif
+}
+
+static int set_thread_event_handler(OSSL_LIB_CTX *ctx, THREAD_EVENT_HANDLER **hands)
+{
+#ifdef FIPS_MODULE
+    return CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_TEVENT_KEY, ctx, hands);
+#else
+    if (destructor_key.sane != -1)
+        return CRYPTO_THREAD_set_local(&destructor_key.value, hands);
+    return 0;
+#endif
+}
+
 static THREAD_EVENT_HANDLER **
-init_get_thread_local(CRYPTO_THREAD_LOCAL *local, int alloc, int keep)
+init_manage_thread_local(OSSL_LIB_CTX *ctx, int alloc, int keep)
 {
-    THREAD_EVENT_HANDLER **hands = CRYPTO_THREAD_get_local(local);
+    THREAD_EVENT_HANDLER **hands = get_thread_event_handler(ctx);
 
     if (alloc) {
         if (hands == NULL) {
 
             if ((hands = OPENSSL_zalloc(sizeof(*hands))) == NULL)
                 return NULL;
 
-            if (!CRYPTO_THREAD_set_local(local, hands)) {
+            if (!set_thread_event_handler(ctx, hands)) {
                 OPENSSL_free(hands);
                 return NULL;
             }
-
 #ifndef FIPS_MODULE
             if (!init_thread_push_handlers(hands)) {
-                CRYPTO_THREAD_set_local(local, NULL);
+                set_thread_event_handler(ctx, NULL);
                 OPENSSL_free(hands);
                 return NULL;
             }
 #endif
         }
     } else if (!keep) {
-        CRYPTO_THREAD_set_local(local, NULL);
+        set_thread_event_handler(ctx, NULL);
     }
 
     return hands;
 }
 
-#ifndef FIPS_MODULE
-/*
- * Since per-thread-specific-data destructors are not universally
- * available, i.e. not on Windows, only below CRYPTO_THREAD_LOCAL key
- * is assumed to have destructor associated. And then an effort is made
- * to call this single destructor on non-pthread platform[s].
- *
- * Initial value is "impossible". It is used as guard value to shortcut
- * destructor for threads terminating before libcrypto is initialized or
- * after it's de-initialized. Access to the key doesn't have to be
- * serialized for the said threads, because they didn't use libcrypto
- * and it doesn't matter if they pick "impossible" or dereference real
- * key value and pull NULL past initialization in the first thread that
- * intends to use libcrypto.
- */
-static union {
-    long sane;
-    CRYPTO_THREAD_LOCAL value;
-} destructor_key = { -1 };
+static ossl_inline THREAD_EVENT_HANDLER **init_fetch_clear_thread_local(OSSL_LIB_CTX *ctx)
```

Comment:
same here: the name `init_fetch_clear`_... it does not help me to understand what's going to happen unless I got and look under the hood.